Add support for merging custom meta data in tribe node

areek opened the pull request:
Currently, when any underlying cluster has custom metadata
(via plugin), tribe node does not store custom meta data in its
cluster state. This is because the tribe node has no idea how to
select the appropriate custom metadata from one or many custom
metadata (corresponding to the number of underlying clusters).

This change adds an interface that custom metadata implementations
can extend to add support for merging mulitple custom metadata of
the same type for storing in the tribe state.

Relates to #20544
Supersedes #20791
Closes #9372

areek commented:
@bleskes can you take a look?

javanna commented:
Seems like this PR fixes #9372 ?

bleskes commented:
FYI @areek is working on allowing custom type instances to be removed as well (which is currently not part of the PR)

areek commented:
@javanna this does implement a solution for #9372, it adds the merge functionality only for `MetaData.Custom`s.
@bleskes I updated the PR according to our discussion offline, now we support removing custom types as well, can you take a look?

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -134,6 +135,28 @@ public static Settings processSettings(Settings settings) {
+    interface MergableCustomMetaData<T extends MetaData.Custom> {
+
+        /**
+         * Merges this custom metadata with other, returning either this or <code>other</code> custom metadata
```
nit: can we document that the neither original customs should be changed by this method?

areek replied:
added it to the docs

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -352,8 +375,7 @@ public String describeTasks(List<ClusterChangedEvent> tasks) {
             try {
-                // we only need to apply the latest cluster state update
-                accumulator = applyUpdate(accumulator, tasks.get(tasks.size() - 1));
+                accumulator = applyUpdate(accumulator, tasks);
```
this method became useless I think? the whole try catch can be removed as it's handled by the caller in the `ClusterService`. I think we should break applyUpdate into several methods each updating a different parts. i.e., updateNodes, updateIndices and now the now updateCustoms

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -463,6 +487,44 @@ private ClusterState applyUpdate(ClusterState currentState, ClusterChangedEvent
+                            final MetaData.Custom custom;
+                            if (currentTribeName.equals(tribeName)) {
+                                // current tribe node
+                                custom = tribeState.metaData().custom(customMetaDataType);
```
Why do you we need the custom of the tribe node?

areek replied:
Apologies for the misleading comment (updated it now). This special cases getting the custom for the tribeClient that triggered the current cluster state update. Because we have the latest cluster state from the `clusterChangedEvent` for this tribeClient, we use that instead of looking up the state from the tribeClient's cluster service. 
I think this is the right thing to do here, though I have tested it out without this special case and it still works. I am worried if the tribeClient's state from the cluster service (which initiated the cluster state update) is outdated w.r.t its latest changes. WDYT @bleskes ?

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -463,6 +487,44 @@ private ClusterState applyUpdate(ClusterState currentState, ClusterChangedEvent
+                                // current tribe node
+                                custom = tribeState.metaData().custom(customMetaDataType);
+                            } else {
+                                ClusterState currentTribeState = tribeNode.injector().getInstance(ClusterService.class).state();
```
nit: can we have this ugliness in a getClusterService(Node node) method? we use it in `doStart` as well.

areek replied:
done

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -463,6 +487,44 @@ private ClusterState applyUpdate(ClusterState currentState, ClusterChangedEvent
+                    .collect(Collectors.toSet());
+            final List<Node> tribeNodes = TribeService.this.nodes;
+            Map<String, MetaData.Custom> mergedCustomMetaDataMap = mergeChangedCustomMetaData(changedCustomMetaDataTypeSet,
+                    customMetaDataType -> {
```
can we make this a proper `getCurrentCustomsFromNodes` method and test it etc?

areek replied:
we only get a list of `MergableCustomMetaData`s from this for all the tribeClient nodes and if we do the special casing as mentioned [here](https://github.com/elastic/elasticsearch/pull/21552#discussion_r88539383), extracting it to a static function and testing it would be difficult. I think we are better off beefing up the integration tests `TribeIT` to deal with multiple customs instead?

areek replied:
I added tribe integration test for merging multiple custom metadata types

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -463,6 +487,44 @@ private ClusterState applyUpdate(ClusterState currentState, ClusterChangedEvent
+                    // custom md has been removed
+                    clusterStateChanged = true;
+                    metaData.removeCustom(changedCustomMetaDataType);
+                } else if (mergedCustomMetaData.equals(tribeCustomMetaData) == false) {
```
do we really need this extra protection? I mean the type was already in our changed type list?

areek replied:
no we don't, I removed the extra protection

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -494,4 +556,30 @@ private void addNewIndex(ClusterState tribeState, ClusterBlocks.Builder blocks,
+            List<MetaData.Custom> tribeCustomMetaDataList = customMetaDataByTribeNode.apply(customMetaDataType);
+            if (tribeCustomMetaDataList.isEmpty() == false) {
+                final MetaData.Custom mergedCustomMetaData;
+                if (tribeCustomMetaDataList.size() == 1) {
```
I don't think you need an if else structure here? you can set the first and iterate over the rest? This can be simplified

areek replied:
I simplified it as suggested

areek commented:
Thanks for the feedback @bleskes! I addressed your comments, left a [question](https://github.com/elastic/elasticsearch/pull/21552#discussion_r88539383) and added more tests. could you take another look?

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -350,21 +374,21 @@ public String describeTasks(List<ClusterChangedEvent> tasks) {
-            } catch (Exception e) {
-                builder.failures(tasks, e);
+            ClusterState.Builder newState = ClusterState.builder(accumulator).incrementVersion();
+            boolean clusterStateChanged = updateNodes(accumulator, tasks, newState);
```
accumulator is not currentstate?

bleskes replied:
thinking about it more - do we even need this accumulator?

areek replied:
I cleaned up the code and removed the redundant accumulator

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -462,13 +495,59 @@ private ClusterState applyUpdate(ClusterState currentState, ClusterChangedEvent
+                        for (Node tribeClientNode : tribeClientNodes) {
+                            String currentTribeName = TRIBE_NAME_SETTING.get(tribeClientNode.settings());
+                            final MetaData.Custom custom;
+                            if (currentTribeName.equals(tribeName)) {
```
Answering here - I would prefer keeping the logic the same for all nodes i.e., any custom change is processed based on the latest cluster state of those nodes. No need to special case the current node as I don't see what it buys us but complexity?  without it it can be a simple stream

bleskes replied:
yes it is - see [here](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java#L777) vs [here](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java#L766).

Might be good to add an assertion the tribe listener to see that the cluster state is already exposed when it's called - this will prevent future trouble :)

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -462,13 +495,59 @@ private ClusterState applyUpdate(ClusterState currentState, ClusterChangedEvent
+                MetaData.Custom mergedCustomMetaData = mergedCustomMetaDataMap.get(changedCustomMetaDataType);
+                if (mergedCustomMetaData == null) {
+                    // custom md has been removed
+                    clusterStateChanged = true;
```
this can be out of if now.

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -494,4 +573,28 @@ private void addNewIndex(ClusterState tribeState, ClusterBlocks.Builder blocks,
+
+    // pkg-private for testing
+    static Map<String, MetaData.Custom> mergeChangedCustomMetaData(Set<String> changedCustomMetaDataTypeSet,
+                                                                   Function<String, List<MetaData.Custom>> customMetaDataByTribeNode) {
```
we can make this a function to `List<MergableCustomMetaData>` - we alreay check with instance of in the implementation of it.

bleskes replied:
if we do so, I think this becomes a simple stream reduce

bleskes commented on core/src/test/java/org/elasticsearch/tribe/TribeServiceTests.java:
```diff
@@ -96,4 +109,128 @@ public void testPassthroughSettings() {
+
+    public void testMergeMultipleCustomMetaData() {
+        Map<String, List<MetaData.Custom>> inputMap = new HashMap<>();
+        inputMap.put(MergableCustomMetaData1.TYPE,
```
I think we need some testing with removed customs? i.e., nulls?

areek replied:
the `mergeChangedCustomMetaData` never gets null values for removed customs, instead they are not supplied by the `customMetaDataByTribeNode` function at all when a mergable custom metadata has been removed. I added tests for removing customs in `TribeIT` (`testMergingRemovedCustomMetaData` and `testMergingMultipleCustomMetaData`)

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -506,34 +501,19 @@ private boolean updateIndicesAndMetaData(ClusterState currentState, List<Cluster
-                        }
-                        return tribeCustomMetaDataList;
-                    }
+                    customMetaDataType -> tribeClientNodes.stream()
```
👍

bleskes commented on core/src/main/java/org/elasticsearch/tribe/TribeService.java:
```diff
@@ -462,13 +490,49 @@ private ClusterState applyUpdate(ClusterState currentState, ClusterChangedEvent
+                MetaData.Custom mergedCustomMetaData = mergedCustomMetaDataMap.get(changedCustomMetaDataType);
+                if (mergedCustomMetaData == null) {
+                    // we ignore merging custom md which doesn't implement MergableCustomMetaData interface
+                    if (currentState.metaData().custom(changedCustomMetaDataType) instanceof MergableCustomMetaData) {
```
I'm wondering if we really need this check - when is it relevant?

areek replied:
when a non-mergable custom md changes, it is in `changedCustomMetaDataTypeSet` but is ignored by `mergeChangedCustomMetaData` (returns a `null` for the custom type). So we can delete non-mergable custom md from the tribe cluster, while trying to merge mergable custom md. This check prevents deleting these non-mergable custom md.

areek commented:
@bleskes I am working on writing some simple unit tests for `TribeNodeClusterStateTaskExecutor`, I plan on testing the bulk processing of incoming states there (by making it to a static class)

areek commented:
I am going to merge this to master and backport it to 5.x branch now. Will add the unit tests (including tsting bulk processing of incoming states) for `TribeNodeClusterStateTaskExecutor` in a subsequent PR.

areek commented:
Thanks for the reviews, @bleskes! merged to master and backported to [5.x](https://github.com/elastic/elasticsearch/commit/57ac65727f45c319e58349ec0cab30cb6a084508)
